core/types/calls.cc

```diff
@@ -1461,36 +1461,54 @@ class T_attached_class : public IntrinsicMethod {
-            return;
+namespace {
+
+enum class MustBehavior {
+    T,
+    Magic,
+};
+
+void implementMust(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res, MustBehavior behavior) {
+    if (args.args.empty()) {
+        return;
+    }
+    const auto loc = args.callLoc();
+    if (!args.args[0]->type.isFullyDefined()) {
+        if (behavior == MustBehavior::Magic) {
+            /* Nothing */
+        } else if (auto e = gs.beginError(loc, errors::Infer::BareTypeUsage)) {
+            e.setHeader("T.must() applied to incomplete type `{}`", args.args[0]->type.show(gs));
         }
-        const auto loc = args.callLoc();
-        if (!args.args[0]->type.isFullyDefined()) {
-            if (auto e = gs.beginError(loc, errors::Infer::BareTypeUsage)) {
-                e.setHeader("T.must() applied to incomplete type `{}`", args.args[0]->type.show(gs));
+        return;
+    }
+    auto ret = Types::approximateSubtract(gs, args.args[0]->type, Types::nilClass());
+    if (ret == args.args[0]->type) {
+        if (behavior == MustBehavior::Magic) {
+            /* Nothing */
+        } else if (auto e = gs.beginError(loc, errors::Infer::InvalidCast)) {
+            if (args.args[0]->type.isUntyped()) {
+                e.setHeader("`{}` called on `{}`, which is redundant", "T.must", args.args[0]->type.show(gs));
+            } else {
+                e.setHeader("`{}` called on `{}`, which is never `{}`", "T.must", args.args[0]->type.show(gs), "nil");
             }
-            return;
-        }
-        auto ret = Types::approximateSubtract(gs, args.args[0]->type, Types::nilClass());
-        if (ret == args.args[0]->type) {
-            if (auto e = gs.beginError(loc, errors::Infer::InvalidCast)) {
-                if (args.args[0]->type.isUntyped()) {
-                    e.setHeader("`{}` called on `{}`, which is redundant", "T.must", args.args[0]->type.show(gs));
-                } else {
-                    e.setHeader("`{}` called on `{}`, which is never `{}`", "T.must", args.args[0]->type.show(gs),
-                                "nil");
-                }
-                e.addErrorSection(args.args[0]->explainGot(gs, args.originForUninitialized));
-                const auto locWithoutTMust = loc.adjust(gs, 7, -1);
-                if (loc.exists() && locWithoutTMust.exists()) {
-                    e.replaceWith("Remove `T.must`", loc, "{}", locWithoutTMust.source(gs).value());
-                }
+            e.addErrorSection(args.args[0]->explainGot(gs, args.originForUninitialized));
+            const auto locWithoutTMust = loc.adjust(gs, 7, -1);
+            if (loc.exists() && locWithoutTMust.exists()) {
+                e.replaceWith("Remove `T.must`", loc, "{}", locWithoutTMust.source(gs).value());
             }
         }
-        res.returnType = move(ret);
+    }
+    if (behavior == MustBehavior::Magic) {
+        ret = Types::approximateSubtract(gs, ret, Types::falseClass());
+    }
```

Comment:
Slight preference for moving this out of `implementMust` and up into `Magic_mustBeTruthy` (i.e., after calling `implementMust`)

Reply:
(continued below)